Sources_v3/StreamChat/Config/Token.swift

```diff
@@ -1,40 +1,41 @@
 //
-// Copyright © 2020 Stream.io Inc. All rights reserved.
+// Copyright © 2021 Stream.io Inc. All rights reserved.
 //
 
 import Foundation
 
-/// A token is used to authenticate a user.
-public typealias Token = String
+enum TokenError: Error {
+    case missingUserId
+}
 
-/// A token provider is a function in which you send a request to your own backend to get a Stream Chat API token.
-///
-/// Set your custom `TokenProvider` in `ChatClientConfig` when creating a `ChatClient` instance. `ChatClient` will use it whenever
-/// it needs to get a new token for the given user.
-///
-public typealias TokenProvider = (_ apiKey: APIKey, _ userId: UserId, _ completion: @escaping (Token?) -> Void) -> Void
+public struct Token: Decodable, Equatable {
+    public let rawValue: String
+    public let userId: UserId
 
-extension Token {
-    /// A token which can be used during development.
-    public static let development: Token = "development"
-    
-    /// Locally checks if the provided token is valid for the given user id.
-    ///
-    /// A token is a string in the JSON Web Token format and one of the parameters it contains is the id of the given user. This
-    /// makes it possible to locally check if the token is valid for the given user.
-    ///
-    /// - Warning: The fact that the token is valid for the given user doesn't mean it can't be rejected by the servers. This is
-    /// a required but not sufficient condition.
-    ///
-    public func isValid(for userId: UserId) -> Bool {
-        if self == .development {
-            return true
+    init(rawValueWithUserID: String) throws {
+        guard let userId = rawValueWithUserID.jwtPayload?["user_id"] as? String else {
+            throw TokenError.missingUserId
         }
-        
-        return !userId.isEmpty && (payload?["user_id"] as? String) == userId
+
+        self.init(rawValue: rawValueWithUserID, userId: userId)
     }
```

Comment:
nit 1: this should be public
nit 2: let's just name it `init(rawValue:)`, the other option is an implementation detail and shouldn't affect the public API in any way.

Reply:
done